### PR TITLE
Address #1566 by adding WithObserver on Queries

### PR DESCRIPTION
**Pull request by kyuff**
This PR adds support for Observability as initially described in #1566.

fixes issue: #1566

### CODE REVIEW

**Comment by kyuff** [comment]
@kyleconroy the PR is updated and should be able to run through CI again. :)

File: examples/authors/mysql/db.go [hunk 1]
```diff
@@ -17,15 +17,28 @@ type DBTX interface {
...
-	db DBTX
+	db       DBTX
+	observer func(ctx context.Context, methodName string) (context.Context, func(err error) error)
+}
+
+func noopObserver(ctx context.Context, methodName string) (context.Context, func(err error) error) {
+	return ctx, func(err error) error { return err }
+}
+
+func (q *Queries) WithObserver(observer func(ctx context.Context, methodName string) (context.Context, func(err error) error)) *Queries {
```
[the remark this answers is not in the document]
Comment (kyuff) [reply examples/authors/mysql/db.go:32]: That's a great input, I'll add a comment with a built-in example shortly. :)

[the remark this answers is not in the document]
Comment (kyuff) [reply examples/authors/mysql/db.go:43]: @kindermoumoute I have added comments inspired by your suggestion.
I didn't link to a doc site as it was unclear where to link to.
Instead the example is inlined in the comment.

**Review by kindermoumoute — APPROVED** [review]
Great! LGTM

**Comment by kyuff** [comment]
@kyleconroy What do you think? 

Is the PR something you want to move on with? And if so, any further changes you see we need?

**Comment by kyleconroy** [comment]
@kyuff Thanks for writing up the original issue and working on an implementation. 

Right now this isn't a feature I'm looking to add to the core of sqlc. I agree that observability is important. I usually suggest that people use [sqlmw](https://github.com/ngrok/sqlmw). I also think that plugins should offer a way to add this without changes to the generated code.

**Comment by kyuff** [comment]
> @kyuff Thanks for writing up the original issue and working on an implementation.
> 
> Right now this isn't a feature I'm looking to add to the core of sqlc. I agree that observability is important. I usually suggest that people use [sqlmw](https://github.com/ngrok/sqlmw). I also think that plugins should offer a way to add this without changes to the generated code.

That's a shame to hear. 

I will look forward to see what we can do with the plugin system once that is in place.

**Pull request closed without merging**